### DISCUSSION TITLE
How to maintain type safety when unwrapping partially loaded entity

### DISCUSSION THREAD

**Discussion by SpencerKaiser** [mikro-orm/mikro-orm#3265] [Q&A]
If I have an entity that has an `IdentifiedReference` to another entity and I load part of that references values, it seems like using `unwrap` nukes the type safety around what was actually loaded

``` typescript
// Book entity
author: IdentifiedReference<Author>;

// ...elsewhere
const book = await em.findOne(
   Book,
   { id: 1 },
   {
     fields: [
       'name',
       'author.name',
     ],
   },
 );

const authorName = book.author.name; // TypeError (expected) because author wasn't unwrapped
const authorName = book.author.unwrap().name; // ✅ 

const author = book.author.unwrap(); // type of `author` is now `Author` instead of `Loaded<Author, 'name'>`
const author = book.author.unwrap() as Loaded<Author, 'name'>; // Workaround to preserve correct types
```

Am I doing something wrong? Is this just a limitation I need to workaround?

**Comment by B4nan** [comment]
`unwrap` is an escape hatch, it just gives you the entity. the intended usage is either ensuring the entity is loaded via `await book.author.load()`, or preferrably via populate hints, which should work out of box in your example, as your `book` should be typed to something like `Loaded<Book, 'author'>` (it will be more verbose but this is the important bit).

With `Loaded` type you have two additional ways to access the entity that work exactly as `unwrap`, but are available only if the populate hint says they are loaded: `$` and `get()` (its an alias, pick one and stick to it). This way you use sync access on things you know are loaded, and this is checked via TS, so on compile time.

```ts
const book = em.findOne(Book, 1, { populate: ['author'] });
const authorName = book.author.$.name; // ✅ 
```

On the other hand, in your example, there is no added value in presering that scalar type hint, as it only works for populating, not for partial loading. This only makes sense if you handle nested population. The populate hint is inferred automatically from partial loading definition, and having the `name` there is just a side effect.

**Reply by SpencerKaiser** [reply]
I'm so sorry for the belated response! I appreciate the context there. Everything there makes perfect sense and it sounds like our entire codebase needs to start using `$`/`get()` in place of `unwrap` because it sounds much more type safe 😅 

The only missing piece from my original post that I think will add valuable context is that once we load an entity and partially load nested fields, we might use that in 5, 10, 15 places and writing out `book.author.$` adds quite a lot of text. Part of the reason I created the reference in my example above is to take `book.author.$` and turn it into `author` to be used everywhere. Would `get()` work to return a reference that has good type safety on what's been loaded?
